content/blog/2023-remote-data-101/index.mdx

```diff
@@ -0,0 +1,145 @@
+* Cosmetic issues: typos in property names, mixing snake case and camel case, misleading naming (field called `distanceInMiles` that returns kilometers instead), etc. Those issues can be fixed on your side, or accepted as quirks and propagated across your codebase as well.
+
+Of course, a lot of those issues are simply differences between how your team sees the domain (and how things _should_ be) and how other teams do. Sometimes there's a business logic error hidden there that you can simply ask the other team to fix, but a lot of times it's a matter of preference. Of course, even if you're right it doesn't mean that the problem is worth fixing. Good luck [raising priority on your ticket that somebody misspelled _referrer_](https://en.wikipedia.org/wiki/HTTP_referer).
+
+If you value maintainability in the long-term over moving fast and breaking things, you'll probably want to draw a line where the data enters your app and put a border guard there that does two things:
+* Checks for correctness
+* Translates from external (backend) to internal (yours) representation (and back to external format when data's leaving your app)
+
+You _could_ hand-craft a layer of artisanal code somewhere between low-level networking and your business logic, but you don't have to. Enter [zod](https://zod.dev/).
+
+## Data validation
+
+Let's take as an example the Marvel API's endpoint for characters. We'd like to represent a list of them with a possibility of fetching more data for a detailed screen should our user tap any of them. We _could_ do manual checks for the fields that we need (`ID` - to later fetch the details, `name` - to display it in our list, `URL` - to fetch a thumbnail image, and `modified` time - to possibly help us with cache invalidation), but we can also let `zod` do it for us by defining a schema that we're interested in:
+
+```tsx
+import z from 'zod'
+
+const HeroUrlBE = z.object({
+  type: z.string(),
+  url: z.string().url(),
+})
+
+const HeroBESchema = z.object({
+  id: z.number(),
+  modified: z.string().datetime({ offset: true }),
+  name: z.string(),
+  urls: z.array(HeroUrlBE).nonempty(),
+})
+```
+
+Then we can simply use this schema to validate data we received using `HeroBESchema.parse(someHeroWannabe)`. 
+
+Note: I often use `BE` sufix to mark types or schemas that represent remote (backend) data representation.
+
+## Data transformation
+
+As I mentioned above, checking correctness is just the first step, because we'd also like to control the shape of data inside our app. 
+
+In that particular case, we get an array of URLs, but we only need one. We can define a schema transformation that will take the remote representation and massage it into shape that we'd like to use. For example:
+
+```tsx
+const HeroFESchema = HeroBESchema.transform(hero => ({
+  url: hero.urls[0].url,
+  ...D.deleteKeys(hero, ['urls']),
```

Comment:
If I didn't know you, I'd be a bit unsure about this line here. I would probably guess that D is something that comes with zod.
Personally I would not use ts-belt in zod examples, but if you want to popularise this tool more, maybe it would be better to mention it below...

Reply:
Good point, I'll add the missing imports, so it's clear 👍